Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,249 @@
+      2. outside the execution of the functions mentioned, this type of access is not possible, because the userspace mappings are made available only while executing such functions
+      3. the userspace memory map can implement HW protections against being misused by the kernel
+      4. the kernel is able to access process pages in the same sequence and with the same mappings as the process does
+   2. through a memory buffer (e.g. a memory area where the kernel regularly needs to perform large amount of read/write operations, like a network buffer)
+      1. the memory used by a buffer is well defined and delimited, rather than a generic area, and it is specifically reserved for this purpose
+      2. the kernel uses own mappings in EL1, while the user space uses the EL0 ones
+      3. very often the kernel doesn't actually access this area directly, but it rather configures a DMA controller to do the transfers, directly to the physical memory.
+      4. misconfiguration of these peripherals that can access physical memory directly is a potential problem - a form of mitigation relies on using IOMMUs: hardware components that the kernel can configure as a firewall on the phisical memory bus, to limit what memory is accessible to each memory bus master device.
+   3. through the EL1 mappings + linear map
+      1. not the intended way for the kernel to access process context, because the intended way is one of the two previously described
+      2. attempting to use EL1 mappings would be not very useful, both for legitimate purposes and even for malicious ones:
+         1. the sequence of pages mapped in the userspace process is not known, and it can change continuously, as the memory management does its job of optimising page allocation to running processes
+         2. some process pages might even be "missing", because they have been either swapped out or dropped (see previous point)
+         3. even for an attacker, in a security scenario, because most likely the attacker would need to access user space pages sequentially, or anyway through user mappings (this is why EL1 mappings are allowed to have access to pages containing EL0 code/data: the security risk is relatively low)
+         4. it bypasses any protection that the process might employ through its own mapping, which, even for legitimate kernel operations on userspace, would be less safe
+6. Assertions about page "mobility", from different perspectives:
+   1. employment of physical pages - observing a physical memory page (e.g. a 4kB chunk aligned on a 4kB boundary) and how it is employed: what sort of content it might host, over time.
+      1. certain physical pages are put to use for a specific purpose, which is (almost) immutable, for the entire duration of the execution of the kernel:
+         1. kernel code from the monolithic portion (except for pages containing kernel init code, which are released after kernel init is completed)
```

Comment:
by monolithic portion you mean code compiled in the Image and not kmod'd later?

Reply:
yes, if here is a better term, I do not know it.

Reply:
I would just say <<kernel code built as part of the boot Image>>, the reason being that the word monolithic is associated with the Kernel address space (on which KOs make no difference....)

Reply:
ok

Reply:
Agreed

Reply:
done